### PR TITLE
Allow newer thecodingmachine/safe

### PR DESCRIPTION
**Pull request by mmoll**
https://github.com/thecodingmachine/safe/pull/322 was needed for the phpstan check to pass.

### CODE REVIEW

**Comment by ricardoboss** [comment]
Are all these changes necessary? I mean wouldnt #1651 suffice?

**Comment by mmoll** [comment]
@ricardoboss If your minimal update does the trick, I'm in favour of it. However, I had some phpstan warnings to fix locally (might be just my machine, though).

**Comment by ricardoboss** [comment]
I noticed you requested to merge into the base branch, whereas my request is for the PHP 8 feature branch (it wouldn't be an immediate update, but it would update `safe` once infection migrates to PHP 8).

**Comment by mmoll** [comment]
should be ready now 💚

File: src/TestFramework/CommandLineBuilder.php [hunk 1]
```diff
@@ -76,8 +77,13 @@ public function build(string $testFrameworkExecutable, array $phpExtraArgs, arra
          *
          * This lets folks use, say, a bash wrapper over phpunit.
          */
-        if ('cli' === PHP_SAPI && $phpExtraArgs === [] && is_executable($testFrameworkExecutable) && shell_exec('command -v php') !== null) {
-            return array_merge([$testFrameworkExecutable], $frameworkArgs);
+        if ('cli' === PHP_SAPI && $phpExtraArgs === [] && is_executable($testFrameworkExecutable)) {
+            try {
+                shell_exec('command -v php');
+
+                return array_merge([$testFrameworkExecutable], $frameworkArgs);
```
Comment (sanmai) [inline src/TestFramework/CommandLineBuilder.php:84]: Alternatively we can use `unsafe_shell_exec` here.
```
use function shell_exec as unsafe_shell_exec;
```
I don't remember what kind of hoops and loops required to make it pass the CI, but if it improves the readability...

File: tests/phpunit/AutoReview/Makefile/MakefileTest.php [hunk 2]
```diff
@@ -62,9 +63,15 @@ final class MakefileTest extends TestCase
...
     {
+        try {
+            shell_exec('command -v timeout');
+            $timeout = 'timeout 2s';
+        } catch (ExecException) {
+            $timeout = '';
+        }
         $output = shell_exec(sprintf(
             '%s make --silent --file %s 2>&1',
-            shell_exec('command -v timeout') !== null ? 'timeout 2s' : '',
```
Comment (sanmai) [inline tests/phpunit/AutoReview/Makefile/MakefileTest.php:67]: And here, but, arguably, the new variant is easier to read.

**Review by sanmai — APPROVED** [review]
LGTM

**Review by maks-rafalko — APPROVED** [review]

File: src/TestFramework/CommandLineBuilder.php (hunk 1 above)
Comment (mmoll) [reply src/TestFramework/CommandLineBuilder.php]: I reverted this and added a suppression to PHPStan.

File: tests/phpunit/AutoReview/Makefile/MakefileTest.php (hunk 2 above)
Comment (mmoll) [reply tests/phpunit/AutoReview/Makefile/MakefileTest.php:67]: I left that one in :)

File: devTools/phpstan-src.neon [hunk 3]
```diff
@@ -47,6 +47,10 @@ parameters:
         -
             path: '../src/Logger/Html/StrykerHtmlReportBuilder.php'
             message: '#return type with generic class ArrayObject does not specify its types\: TKey, TValue#'
+        -
+            message: "#^Function shell_exec is unsafe to use\\. It can return FALSE instead of throwing an exception\\. Please add 'use function Safe\\\\shell_exec;' at the beginning of the file to use the variant provided by the 'thecodingmachine/safe' library\\.$#"
```
Comment (sanmai) [inline devTools/phpstan-src.neon:51]:
```suggestion
            message: "#^Function shell_exec is unsafe to use#"
```
Nobody wants a build to fail if a single letter changes in that long explanation.

**Review by sanmai — APPROVED** [review]

**Comment by sanmai** [comment]
@mmoll I'd appreciate if you don't force-push. Makes reviewing new changes a lot harder :cry:

**Review by sanmai — APPROVED** [review]
Still good.

**Comment by maks-rafalko** [comment]
Thank you @mmoll & @sanmai !

**Pull request merged**